---
title: "Differential Rhythmicity Paper"
author: "Bharath Ananthasubramaniam"
date: "09/03/2020"
output: html_document
---

```{r packages, echo=FALSE}
library(compareRhythms)
suppressPackageStartupMessages(library(genefilter))
suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(rain))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(DODR))
suppressPackageStartupMessages(library(patchwork))
library(parallel)
library(ggthemes)
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(clusterProfiler))
library(msigdbr)
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(org.Mm.eg.db))
library(openxlsx)
library(precrec)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, include = TRUE, warning = FALSE, fig.height = 3, fig.path='figures/',
                      dev = c('pdf', 'svglite'))
amp_choices <- c(0, 0.5)
set.seed(1234)

m_df <- msigdbr(species = "Mus musculus", category = "H")
m_t2g_hallmark <- m_df %>% dplyr::select(gs_name, gene_symbol) %>% as.data.frame()

m_df <- msigdbr(species = "Mus musculus", category = "C3", subcategory = "TFT:GTRD")
m_t2g_tft <- m_df %>% dplyr::select(gs_name, gene_symbol) %>% as.data.frame()

m_df <- msigdbr(species = "Mus musculus", category = "C2", subcategory = "CP:REACTOME")
m_t2g_reactome <- m_df %>% dplyr::select(gs_name, gene_symbol) %>% as.data.frame()
```

```{r custom_theme}
theme_custom <- function(base_size = 11, base_family = "Helvetica") {
  theme_foundation(base_size = base_size, base_family = base_family) +
    theme(
      line = element_line(size = base_size/11, color = "grey20"),
      rect = element_rect(fill = "white", size = base_size/15),
      text = element_text(size = base_size),
      plot.margin = margin(0.1, 0.1, 0.1, 0.1, unit = "lines"),
      plot.background = element_rect(linetype = 0),
      
      panel.border = element_blank(),
      panel.background = element_blank(),
      
      axis.line = element_line(size = rel(0.6)),
      axis.text = element_text(size = rel(0.95)),
      axis.text.x.bottom = element_text(vjust = 0.25),
      axis.text.y.right = element_text(hjust = 0.25),

      axis.ticks = element_line(size = rel(0.6)),
      axis.ticks.length = unit(0.4, "lines"),
      axis.title = element_text(size = rel(1)),
      axis.title.x.bottom = element_text(vjust = 0),
      axis.title.y.left = element_text(vjust = 1),
      
      legend.background = element_rect(),
      legend.position = "top",
      legend.direction = "horizontal",
      legend.box = "horizontal",
      legend.title = element_text(size = rel(0.95)),
      legend.text = element_text(size = rel(0.95)),
      legend.box.spacing = unit(0.25, "lines"),
      legend.key = element_rect(color = NA),
      legend.key.size = unit(1.5, "lines"),
      legend.box.background = element_blank(),
      legend.margin = margin(t = 0, unit = "lines"),
      
      panel.grid = element_line(size = rel(0.33)),
      panel.grid.major = element_line(color = "grey80"),
      panel.grid.minor = element_blank(),

      strip.text = element_text(size = rel(0.95)),
      strip.background = element_blank(),
      panel.spacing = unit(1, "lines"),
      
      plot.tag = element_text(face = "bold"),
      plot.caption = element_blank(),
      plot.title = element_blank(),
      plot.subtitle = element_blank(),

      aspect.ratio = 0.75
    )
}
theme_set(theme_custom(base_size = 9))
update_geom_defaults("line", list(size = 0.8))
```


```{r functions}
rhythm_analysis <- function(y, exp_design, period = 24) {

  exp_design <- base::cbind(exp_design,
                            col_number = base::seq(base::nrow(exp_design)))

  group_id <- base::unique(exp_design$group)
  assertthat::are_equal(length(group_id), 2)

  exp_design_A <- exp_design[exp_design$group == group_id[1], ]

  exp_design_A <- exp_design_A[base::order(exp_design_A$time), ]

  exp_design_B <- exp_design[exp_design$group == group_id[2], ]

  exp_design_B <- exp_design_B[base::order(exp_design_B$time), ]

  deltat_A <- min(diff(base::unique(exp_design_A$time)))

  time_A <- base::seq(min(exp_design_A$time),
                      max(exp_design_A$time),
                      by = deltat_A)

  unique_times_A <- table(exp_design_A$time)

  measure_sequence_A <- vapply(time_A,
                                     function(t) {
                                       ifelse(any(names(unique_times_A) == t),
                                              unique_times_A[names(unique_times_A) == t],
                                              0)
                                     },
                                     integer(1))


  deltat_B <- min(diff(unique(exp_design_B$time)))

  time_B <- seq(min(exp_design_B$time),
                      max(exp_design_B$time),
                      by = deltat_B)

  unique_times_B <- table(exp_design_B$time)

  measure_sequence_B <- vapply(time_B,
                               function(t) {
                                 ifelse(any(names(unique_times_B) == t),
                                        unique_times_B[names(unique_times_B) == t],
                                        0)
                               },
                               integer(1))

  y_A <- y[, exp_design_A$col_number]
  y_B <- y[, exp_design_B$col_number]

  assertthat::assert_that((sum(measure_sequence_A) >= 12) && (sum(measure_sequence_B) >= 12),
                          msg = "Not enough samples to run RAIN rhythmicity analysis confidently.")

  rain_A <- mcparallel(rain(t(y_A), deltat_A, period,
                       measure.sequence = measure_sequence_A), "A")

  rain_B <- mcparallel(rain(t(y_B), deltat_B, period,
                       measure.sequence = measure_sequence_B), "B")
  rain_analysis <- mccollect(list(rain_A, rain_B))

  circ_params_A <- compute_circ_params(y_A, exp_design_A$time, period = period)

  circ_params_B <- compute_circ_params(y_B, exp_design_B$time, period = period)
  
  rain_results <- data.frame(adj_p_val_A = p.adjust(rain_analysis[[1]][, "pVal"], method = "BH"),
                             adj_p_val_B = p.adjust(rain_analysis[[2]][, "pVal"], method = "BH"),
                             A_amp = circ_params_A,
                             B_amp = circ_params_B)
   
  rownames(rain_results) <- rownames(y)
  return(rain_results)
}

compute_circ_params <- function(y, t, period) {

  inphase <- cos(2 * pi * t / period)
  outphase <- sin(2 * pi * t / period)

  X <- model.matrix(~inphase + outphase)

  fit <- lm.fit(X, t(y))

  amps <- 2 * sqrt(base::colSums(fit$coefficients[-1, ]^2))

  return(amps)

}

diff_rhythmicity_analysis <- function(y, exp_design, results, rhythm_fdr = 0.05, amp_cutoff = 0, period = 24) {
  rhythmic_in_either <- (results$adj_p_val_A < rhythm_fdr & results$A_amp > amp_cutoff) | 
                          (results$adj_p_val_B < rhythm_fdr & results$B_amp > amp_cutoff)
  
  group_id <- base::unique(exp_design$group)

  dodr_results <- robustDODR(t(y[rhythmic_in_either, exp_design$group == group_id[1]]), 
                       t(y[rhythmic_in_either, exp_design$group == group_id[2]]),
                       exp_design$time[exp_design$group == group_id[1]],
                       exp_design$time[exp_design$group == group_id[2]],
                       norm = FALSE, 
                       period = period)
  dodr_results$adj_p_val <- stats::p.adjust(dodr_results$p.value, method = "BH")
  
  return(dodr_results)
}

filter_genes <- function(y, group, min.count=10, min.expr=4, type="ma", n.prop=0.7) {
  stopifnot(type %in% c("ma", "rnaseq"))
  yA <- y[, group == levels(group)[1]]
  yB <- y[, group == levels(group)[2]]
  if (type == "ma") {
    keep <- pmin(rowMeans(yA>min.expr), rowMeans(yB>min.expr)) > n.prop
  } else {
    medianLibSize <- median(colSums(y))
    cpm.cutoff <- min.count/medianLibSize*1e6
    keep <- pmin(rowMeans(edgeR::cpm(yA)>cpm.cutoff), 
                 rowMeans(edgeR::cpm(yB)>cpm.cutoff)) > n.prop
  }
  return(keep)
}

shifter <- function(x, n = 1) {
  if (n == 0) x else c(tail(x, -n), head(x, n))
}

prc_interpolate <- function(TP, FP, maxTrue, maxFeatures) {
  x <- seq(maxTrue-TP)
  return(data.frame("TP" = TP + x, "FP" = FP + (maxFeatures - maxTrue - FP)*x/(maxTrue - TP), "FN" = maxTrue-TP-x))
}
```

## Problem
We are interested in comparing rhythms across two different conditions or treatments. In particular, we would like to know how many time series are differential rhythmic and what are these species. We are using a genomic dataset, but ideas apply to comparison of any two datasets.

## Revealing examples using liver data

* Liver dataset from Hughes et al. was sampled every 1h over 2 days -- highest resolution available
* We create a pair of artificial time series datasets by splitting the data into 2
    + collect *even* and *odd* time points to form two datasets samples every 2h over 2 days each
* Test this pair of datasets of differential rhythmic genes using the common approach
* Due to our construction, we expect no DR genes in this set. 

```{r load-hughes-data}
liver_hughes <- readRDS("data/GSE11923.RDS")

f1 <- kOverA(48, 5)
sel <- genefilter(assay(liver_hughes), f1)

liver_hughes <- liver_hughes[sel, ]

## ----------create artificial group of odd and even samples -------##
exp_design <- as.data.frame(colData(liver_hughes))
exp_design$time <- as.numeric(do.call(rbind, strsplit(as.character(liver_hughes$title), " ", fixed = TRUE))[,3])
exp_design$group <- factor(ifelse(exp_design$time %% 2, "O", "E"))
```

```{r analyze-odd-even, fig.width=4, eval=TRUE}
pvm_ex_1 <- rhythm_analysis(assay(liver_hughes), exp_design)

df1_1 <- expand.grid(fdr_cutoff = 10^seq(-3.5, 0, by = 0.2),
                  amp_cutoff = amp_choices)
df1_1$no_of_DR <- vapply(seq(nrow(df1_1)),
                   function(i) {
                     return(sum(xor(pvm_ex_1$adj_p_val_A < df1_1[i, "fdr_cutoff"] &
                                    pvm_ex_1$A_amp > df1_1[i, "amp_cutoff"],
                                    pvm_ex_1$adj_p_val_B < df1_1[i, "fdr_cutoff"] &
                                    pvm_ex_1$B_amp > df1_1[i, "amp_cutoff"])))
                   }, FUN.VALUE = numeric(1))
df1_1$percent <- df1_1$no_of_DR/nrow(liver_hughes)

f1_1 <- ggplot(df1_1) + 
  geom_line(aes_(x=~fdr_cutoff, y=~no_of_DR, linetype=~factor(amp_cutoff))) +
  scale_linetype_manual(name = "", labels=parse(text = paste0('A > ', amp_choices)), 
                        values = c("solid", "dashed")) +
  xlab("false discovery cutoff for rhythms") +
  ylab("DiffR genes") + coord_cartesian(xlim = c(0.0005,0.2), ylim=c(0, 3000)) +
  scale_x_continuous(trans = "log10", breaks = c(0.001, 0.01, 0.05, 0.2)) +
  annotate("line", x = c(10^-3.4, 10^-2.25), y = 2750) +
  annotate("point", x = 10^(seq(-3.25, -2.5, by = 0.25)), y = 2750, size = 3.5, color="grey20") +
  annotate("point", x = 10^(seq(-3, -2.5, by = 0.5)), y = 2750, shape = "O", color = "white", size = 1.75) +
  annotate("point", x = 10^(seq(-3.25, -2.25, by = 0.5)), y = 2750, shape = "E", color = "white", size = 1.75) +
  theme(legend.box.margin=margin(b=-20), legend.margin = margin())

mod_sel_ex_1 <- compareRhythms(assay(liver_hughes), exp_design = exp_design, method = "mod_sel")
dodr_ex_1 <- compareRhythms(assay(liver_hughes), exp_design = exp_design, method = "dodr")
limma_ex_1 <- compareRhythms(assay(liver_hughes), exp_design = exp_design, method = "limma")

```

* for a range of FDR cutoffs for the rhythmicity in each dataset, anywhere from 10-20% DR genes are found
* these are all false positives as there are no DR genes in this set
* for the typical 0.05 choice of cutoff, the standard method yields 25% false positives
* A variant of this method, also uses an amplitude threshold to enforce biological relevance.
* The amplitude threshold does drastically reduce the number of false positive, but still not eliminated.

* Downsampled liver time series with a sample every 2h also shows large number of false positives
* But the a certain level of false positives occurs at larger FDR cutoffs -- since p-values are larger in the lower resolution data.
```{r artificial-DR-dataset, eval=TRUE}
##--------create artificial DR genes by exchanging labels of genes-------##
N <- 500 #round(0.1*sum(pmin(pvm_ex_1$adj_p_val_A, pvm_ex_1$adj_p_val_B)<0.05))

amp_cutoff <- median(pvm_ex_1$A_amp[pvm_ex_1$adj_p_val_A<0.05])

large_amp_genes <- pvm_ex_1[pvm_ex_1$A_amp > amp_cutoff & pvm_ex_1$adj_p_val_A<0.05, ]

rhythmic_in_A <- rownames(large_amp_genes)[sample.int(nrow(large_amp_genes))][seq(min(N, nrow(large_amp_genes)))]

rhythmic_in_A_large_amp <- rhythmic_in_A[pvm_ex_1[rhythmic_in_A, "A_amp"] > 0.5]

new_expr <- assay(liver_hughes)

for (gene in rhythmic_in_A) {
  temp <- new_expr[gene, ]
  scale_shift = runif(1, 0.5, 1.5)*(temp[exp_design$group == "O"] - mean(temp[exp_design$group == "O"])) + mean(temp[exp_design$group == "O"])
  temp[exp_design$group == "O"] <- shifter(scale_shift, sample(length(scale_shift), 1))
  new_expr[gene, ] <- temp
}
```

```{r Figure1, fig.width = 5, fig.height=4.5, eval=TRUE}
pvm_ex_2 <- rhythm_analysis(new_expr, exp_design)

df1_2 <- expand.grid(fdr_cutoff = 10^seq(-9, log10(0.999), by=0.1),
                  amp_cutoff = amp_choices)
true_dr_genes <- rownames(pvm_ex_2) %in% rhythmic_in_A

roc <- vapply(seq(nrow(df1_2)),
              function(i) {
                est_dr_genes <- xor(pvm_ex_2$adj_p_val_A < df1_2[i, "fdr_cutoff"] &
                                      pvm_ex_2$A_amp > df1_2[i, "amp_cutoff"],
                                    pvm_ex_2$adj_p_val_B < df1_2[i, "fdr_cutoff"] &
                                      pvm_ex_2$B_amp > df1_2[i, "amp_cutoff"])
                tp <- sum(true_dr_genes & est_dr_genes)
                fn <- sum(true_dr_genes & !est_dr_genes)
                fp <- sum(!true_dr_genes & est_dr_genes)
                tn <- sum(!true_dr_genes & !est_dr_genes)
                return(c(TP = tp, FP = fp, FN = fn, TN=tn))
              }, FUN.VALUE = numeric(4))
df1_2 <- cbind(df1_2, t(roc))

f1_2 <- ggplot(df1_2) +
  geom_area(aes_(x=~fdr_cutoff, y=~FP+TP), color = "black", fill = "white") +
  geom_area(aes_(x=~fdr_cutoff, y=~TP), fill = "grey70", position="identity", alpha=0.75) +
  facet_wrap(~factor(amp_cutoff), ncol = 1,
             labeller = as_labeller(setNames(paste0('~A>', amp_choices), amp_choices), label_parsed), scales = "free") +
  geom_hline(aes_(yintercept=~length(rhythmic_in_A)), linetype = "twodash", size = 0.6)  +
  xlab("false discovery cutoff for rhythms") + ylab("DiffR transcripts") +
  scale_x_continuous(trans = "log10", expand = c(0,0), breaks = c(0.0001, 0.001, 0.01, 0.1),
                     labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  coord_cartesian(xlim = c(0.00001,0.5)) + theme(panel.spacing.y = unit(1.2, "lines"))

point_1 <- with(df1_2, which.min(abs(TP/(TP+FP)-0.2) + abs(TP/(TP+FN)-0.2)))
point_2 <- with(df1_2, which.min(abs(fdr_cutoff - 0.05)))

f1_3 <- ggplot(df1_2, aes_(x=~TP/(TP+FN), y=~TP/(FP+TP), linetype=~factor(amp_cutoff))) + 
  geom_hline(aes_(yintercept=~N/(N + nrow(pvm_ex_2))), color="grey50", linetype="twodash") +
  geom_path(color="black") + scale_linetype_manual(name="", labels=parse(text = paste0('A > ', amp_choices)), 
                        values = c("solid", "dashed")) + 
  geom_point(aes_(x=~TP/(TP+FN), y=~TP/(FP+TP)), shape=21, size=2, stroke=0.75, fill="white", data = df1_2[point_1,]) +
  geom_point(aes_(x=~TP/(TP+FN), y=~TP/(FP+TP)), shape=21, size=2, stroke=0.75, fill="grey50", data = df1_2[point_2,]) +
  xlab("Recall = DiffR found/DiffR present") + 
  ylab("Precision = true DiffR hits/ DiffR hits") + coord_cartesian(xlim=c(0,0.5), ylim=c(0,0.5))


layout <- c(area(1,1), area(1,2,2,2), area(2,1))


f1_1 + f1_2 + f1_3 + plot_layout(design = layout, widths = c(1,0.9)) + plot_annotation(tag_levels = 'A')

```

```{r pvm_roc_a_cutoff, eval=FALSE}
df1_2_2 <- expand.grid(fdr_cutoff = 10^seq(-9, log10(0.999), by=0.1),
                  amp_cutoff = amp_choices)

true_dr_genes <- rownames(pvm_ex_2) %in% rhythmic_in_A_large_amp

roc <- vapply(seq(nrow(df1_2_2)),
              function(i) {
                est_dr_genes <- xor(pvm_ex_2$adj_p_val_A < df1_2[i, "fdr_cutoff"] &
                                      pvm_ex_2$A_amp > df1_2[i, "amp_cutoff"],
                                    pvm_ex_2$adj_p_val_B < df1_2[i, "fdr_cutoff"] &
                                      pvm_ex_2$B_amp > df1_2[i, "amp_cutoff"])
                tp <- sum(true_dr_genes & est_dr_genes)
                fn <- sum(true_dr_genes & !est_dr_genes)
                fp <- sum(!true_dr_genes & est_dr_genes)
                tn <- sum(!true_dr_genes & !est_dr_genes)
                return(c(TP = tp, FP = fp, FN = fn, TN=tn))
              }, FUN.VALUE = numeric(4))
df1_2_2 <- cbind(df1_2_2, t(roc))

ggplot(df1_2_2, aes_(x=~TP/(TP+FN), y=~TP/(FP+TP), linetype=~factor(amp_cutoff))) + 
  geom_hline(aes_(yintercept=~N/(N + nrow(pvm_ex_2))), color="grey90") +
  geom_path(color="black") + scale_linetype_manual(name="", labels=parse(text = paste0('A > ', amp_choices)), 
                        values = c("solid", "dashed")) + 
  xlab("Recall = DiffR found/DiffR present") + 
  ylab("Precision = true DiffR hits/ DiffR hits") + coord_cartesian(xlim=c(0,0.5), ylim=c(0,0.5))
```


* for all typical choices of fdr cutoff, larger number of DR genes are predicted by PVM
* However, the predicted DR genes account for only for 80% of true DR genes
* Large numbers of false DR (false positives) genes
* Notice the number of true DR genes in set varies non-monotonically.
* Using an amplitude cutoff, controls the inflation in the number of DR genes, but at the cost of even fewer true hits (lower power)

## Applying the appropriate methods

```{r Figure3, fig.width=5, fig.height=3}
all_genes <- data.frame(id = rownames(new_expr))

limma_ex_2_1 <- compareRhythms(new_expr, exp_design = exp_design, method = "limma",
                             compare_fdr = 1.0, amp_cutoff = amp_choices[1], just_classify = FALSE) %>%
                right_join(all_genes, by = "id") %>% arrange(id)

limma_ex_2_2 <- compareRhythms(new_expr, exp_design = exp_design, method = "limma",
                             compare_fdr = 1.0, amp_cutoff = amp_choices[2], just_classify = FALSE) %>%
                right_join(all_genes, by = "id") %>% arrange(id)

df3_1 <- evalmod(scores = join_scores(-log10(limma_ex_2_1$adj_p_val_DR),
                                      -log10(limma_ex_2_2$adj_p_val_DR)),
        labels = limma_ex_2_1$id %in% rhythmic_in_A,
        modnames = as.character(amp_choices)) %>%
        fortify(reduce_points=TRUE) %>% filter(curvetype == "PRC") %>%
        mutate(method = "limma", dset = "low", amp_cutoff = as.numeric(levels(modname))[modname])

df3_2 <- evalmod(scores = join_scores(-log10(limma_ex_2_1$adj_p_val_DR),
                                      -log10(limma_ex_2_2$adj_p_val_DR)),
        labels = limma_ex_2_1$id %in% rhythmic_in_A_large_amp,
        modnames = as.character(amp_choices)) %>%
        fortify(reduce_points=TRUE) %>% filter(curvetype == "PRC") %>%
        mutate(method = "limma", dset = "high", amp_cutoff = as.numeric(levels(modname))[modname])

dodr_ex_2_1 <- compareRhythms(new_expr, exp_design = exp_design, method = "dodr",
                             compare_fdr = 1.0, amp_cutoff = amp_choices[1], just_classify = FALSE) %>%
                right_join(all_genes, by = "id") %>% arrange(id)

dodr_ex_2_2 <- compareRhythms(new_expr, exp_design = exp_design, method = "dodr",
                             compare_fdr = 1.0, amp_cutoff = amp_choices[2], just_classify = FALSE) %>%
                right_join(all_genes, by = "id") %>% arrange(id)

df3_3 <- evalmod(scores = join_scores(-log10(dodr_ex_2_1$adj_p_val_dodr),
                                      -log10(dodr_ex_2_2$adj_p_val_dodr)),
        labels = dodr_ex_2_1$id %in% rhythmic_in_A,
        modnames = as.character(amp_choices)) %>%
        fortify(reduce_points=TRUE) %>% filter(curvetype == "PRC") %>%
        mutate(method = "DODR", dset = "low", amp_cutoff = as.numeric(levels(modname))[modname])

df3_4 <- evalmod(scores = join_scores(-log10(dodr_ex_2_1$adj_p_val_dodr),
                                      -log10(dodr_ex_2_2$adj_p_val_dodr)),
        labels = dodr_ex_2_1$id %in% rhythmic_in_A_large_amp,
        modnames = as.character(amp_choices)) %>%
        fortify(reduce_points=TRUE) %>% filter(curvetype == "PRC")%>%
        mutate(method = "DODR", dset = "high", amp_cutoff = as.numeric(levels(modname))[modname])

mod_sel_ex_2_1 <- compareRhythms(new_expr, exp_design = exp_design, method = "mod_sel",
                             schwarz_wt_cutoff = 0.0, amp_cutoff = amp_choices[1], just_classify = FALSE) %>%
                right_join(all_genes, by = "id") %>% arrange(id)

mod_sel_ex_2_2 <- compareRhythms(new_expr, exp_design = exp_design, method = "mod_sel",
                             schwarz_wt_cutoff = 0.0, amp_cutoff = amp_choices[2], just_classify = FALSE) %>%
                right_join(all_genes, by = "id") %>% arrange(id)

grid_points <- sort(mod_sel_ex_2_1$weights)
grid_points <- grid_points[seq(1, length(grid_points), 20)]

df3_5 <- expand.grid(schwarz_wt_cutoff = grid_points,
                  amp_cutoff = amp_choices)

roc <- vapply(seq(nrow(df3_5)),
              function(i) {
                if (df3_5[i, "amp_cutoff"] == amp_choices[1]) {
                    est_dr_genes <- filter(mod_sel_ex_2_1, weights > df3_5[i, "schwarz_wt_cutoff"] & category %in% c("loss", "gain", "change"))
                } else {
                    est_dr_genes <- filter(mod_sel_ex_2_2, weights > df3_5[i, "schwarz_wt_cutoff"] & category %in% c("loss", "gain", "change"))
                }
                tp <- length(intersect(est_dr_genes$id, rhythmic_in_A))
                fn <- length(setdiff(rhythmic_in_A, est_dr_genes$id))
                fp <- length(setdiff(est_dr_genes$id, rhythmic_in_A))
                return(c(TP = tp, FP = fp, FN = fn))
              }, FUN.VALUE = numeric(3))
df3_5 <- cbind(df3_5, t(roc))

df3_5ext <- df3_5 %>% group_by(amp_cutoff) %>% 
  arrange(desc(TP)) %>% 
  filter(row_number()==1) %>%
  summarize(extn = prc_interpolate(TP, FP, 500, nrow(all_genes)), .groups="rowwise") %>%
  mutate(schwarz_wt_cutoff = 0, .before=amp_cutoff) %>%
  as.matrix()

colnames(df3_5ext) <- gsub("extn.", "", colnames(df3_5ext))
df3_5 <- bind_rows(df3_5, data.frame(df3_5ext)) %>%
         group_by(amp_cutoff) %>% arrange(desc(TP)) %>%
        mutate(method = "model sel.", 
               dset = "low",
               x = TP/(TP + FN),
               y = TP/(TP + FP))

df3_6 <- expand.grid(schwarz_wt_cutoff = grid_points,
                  amp_cutoff = amp_choices)

roc <- vapply(seq(nrow(df3_6)),
              function(i) {
                if (df3_6[i, "amp_cutoff"] == amp_choices[1]) {
                    est_dr_genes <- filter(mod_sel_ex_2_1, weights > df3_6[i, "schwarz_wt_cutoff"] & category %in% c("loss", "gain", "change"))
                } else {
                    est_dr_genes <- filter(mod_sel_ex_2_2, weights > df3_6[i, "schwarz_wt_cutoff"] & category %in% c("loss", "gain", "change"))
                }
                tp <- length(intersect(est_dr_genes$id, rhythmic_in_A_large_amp))
                fn <- length(setdiff(rhythmic_in_A_large_amp, est_dr_genes$id))
                fp <- length(setdiff(est_dr_genes$id, rhythmic_in_A_large_amp))
                return(c(TP = tp, FP = fp, FN = fn))
              }, FUN.VALUE = numeric(3))
df3_6 <- cbind(df3_6, t(roc))

df3_6ext <- df3_6 %>% group_by(amp_cutoff) %>% 
  arrange(desc(TP)) %>% 
  filter(row_number()==1) %>%
  summarize(extn = prc_interpolate(TP, FP, length(rhythmic_in_A_large_amp), nrow(all_genes)), .groups="rowwise") %>%
  mutate(schwarz_wt_cutoff = 0, .before=amp_cutoff) %>%
  as.matrix()

colnames(df3_6ext) <- gsub("extn.", "", colnames(df3_6ext))
df3_6 <- bind_rows(df3_6, data.frame(df3_6ext)) %>%
         group_by(amp_cutoff) %>% arrange(desc(TP)) %>%
        mutate(method = "model sel.", 
               dset = "high",
               x = TP/(TP + FN),
               y = TP/(TP + FP))

df3 <- bind_rows(df3_1, df3_2, df3_3, df3_4, df3_5, df3_6) %>%
       dplyr::select(c("x", "y", "amp_cutoff", "method", "dset"))

op_points <- lapply(list(limma_ex_2_2, dodr_ex_2_2, mod_sel_ex_2_2),
       function(d) {
         if ("weights" %in% colnames(d)) {
           est_dr_genes <- filter(d, weights>0.6 & category %in% c("gain", "loss", "change"))
           method <- "model sel."
         } else if ("adj_p_val_DR" %in% colnames(d)) {
           est_dr_genes <- filter(d, adj_p_val_DR<0.05)
           method <- "limma"
         } else {
           est_dr_genes <- filter(d, adj_p_val_dodr<0.05)
           method <- "DODR"
         }
          tp <- c(length(intersect(rhythmic_in_A, est_dr_genes$id)),
                  length(intersect(rhythmic_in_A_large_amp, est_dr_genes$id)))
          fn <- c(length(setdiff(rhythmic_in_A, est_dr_genes$id)),
                  length(setdiff(rhythmic_in_A_large_amp, est_dr_genes$id)))
          fp <- c(length(setdiff(est_dr_genes$id, rhythmic_in_A)),
                  length(setdiff(est_dr_genes$id, rhythmic_in_A_large_amp)))
          x <- tp/(tp + fn)
          y <- tp/(tp + fp)
          return(data.frame(x=x, y=y, amp_cutoff=0.5, method = method, dset=c("low","high")))
       }) %>% 
  bind_rows

ggplot(df3, aes(x=x, y = y, color=method, linetype=factor(amp_cutoff))) + geom_line(size=0.5) + 
  geom_point(shape=21, fill="white", data=op_points, stroke=0.75, show.legend = FALSE) +
  facet_grid(cols = vars(factor(dset, levels = c("low", "high"))), labeller = as_labeller(c(low = "All transcripts", high = "Only large amplitude transcripts"))) + 
  scale_linetype_manual(name="", labels=parse(text = paste0('A > ', amp_choices)), 
                        values = c("solid", "dashed")) + scale_color_colorblind(name="") +
  xlab("Recall = DiffR found/DiffR present") + 
  ylab("Precision = true DiffR hits/ DiffR hits")
```

```{r hughes-data-supp, eval=FALSE}
liver_hughes_sub <- liver_hughes[, exp_design$time %% 2 == 0]
exp_design_sub <- exp_design[exp_design$time %% 2 == 0, ]
exp_design_sub$group <- ifelse(exp_design_sub$time %% 4, "O", "E")
```

```{r analyze-odd-even-supp, fig.width=4, eval=FALSE}
pvm_ex_S1 <- rhythm_analysis(assay(liver_hughes_sub), exp_design_sub)

dfS1_1 <- expand.grid(fdr_cutoff = 10^seq(-3.5, 0, by = 0.2),
                  amp_cutoff = amp_choices)
dfS1_1$no_of_DR <- vapply(seq(nrow(dfS1_1)),
                   function(i) {
                     return(sum(xor(pvm_ex_S1$adj_p_val_A < dfS1_1[i, "fdr_cutoff"] &
                                    pvm_ex_S1$A_amp > dfS1_1[i, "amp_cutoff"],
                                    pvm_ex_S1$adj_p_val_B < dfS1_1[i, "fdr_cutoff"] &
                                    pvm_ex_S1$B_amp > dfS1_1[i, "amp_cutoff"])))
                   }, FUN.VALUE = numeric(1))
dfS1_1$percent <- dfS1_1$no_of_DR/nrow(liver_hughes)

fS1_1 <- ggplot(dfS1_1) + 
  geom_line(aes_(x=~fdr_cutoff, y=~percent, linetype=~factor(amp_cutoff))) +
  theme(legend.position = "top", legend.margin=margin(0,0,0,0)) + 
  scale_linetype_manual(name = "", labels=parse(text = paste0('A > ', amp_choices)), 
                        values = c("solid", "dashed")) +
  scale_y_continuous(labels = scales::percent) + xlab("false discovery cutoff for rhythms") +
  ylab("DiffR genes") + coord_cartesian(xlim = c(0.005,0.25)) + 
  scale_x_continuous(trans = "log10", breaks = c(0.01, 0.05, 0.2)) +
    annotate("line", x = c(10^-2.3, 10^-1.35), y = 0.32) +
  annotate("point", x = 10^(seq(-2.2, -1.4, by = 0.15)), y = 0.32, size = 3.5, color = "grey20") +
    annotate("point", x = 10^(seq(-2.05, -1.4, by = 0.3)), y = 0.32, size = 3, color = "grey90") +
  annotate("point", x = 10^(seq(-2.2, -1.4, by = 0.6)), y = 0.32, shape = "E", color = "white", size = 1.75) +
    annotate("point", x = 10^(seq(-1.9, -1.4, by = 0.6)), y = 0.32, shape = "O", color = "white", size = 1.75) +
      theme(legend.box.margin=margin(b=-20), legend.margin = margin())

```

```{r artificial-DR-data-supp, eval=FALSE}
#N <- 5000.1*nrow(pvm_ex_S1)

large_amp_genes <- pvm_ex_S1[pvm_ex_S1$A_amp > amp_cutoff, ]

rhythmic_in_A <- rownames(large_amp_genes)[order(large_amp_genes$adj_p_val_A)[seq(min(N, nrow(large_amp_genes)))]]

new_expr_sub <- assay(liver_hughes_sub)

for (gene in rhythmic_in_A) {
  temp <- new_expr_sub[gene, ]
  temp[exp_design_sub$group == "O"] <- sample(temp[exp_design_sub$group == "O"])
  new_expr_sub[gene, ] <- temp
}
```

```{r FigureS1, fig.width=7, fig.height=2.5, eval=FALSE}
pvm_ex_S2 <- rhythm_analysis(new_expr_sub, exp_design_sub)

dfS1_2 <- expand.grid(fdr_cutoff = 10^seq(-6, log10(0.25), by=0.05),
                  amp_cutoff = amp_choices)
true_dr_genes <- rownames(pvm_ex_S2) %in% rhythmic_in_A

roc <- vapply(seq(nrow(dfS1_2)),
              function(i) {
                est_dr_genes <- xor(pvm_ex_S2$adj_p_val_A < dfS1_2[i, "fdr_cutoff"] &
                                      pvm_ex_S2$A_amp > dfS1_2[i, "amp_cutoff"],
                                    pvm_ex_S2$adj_p_val_B < dfS1_2[i, "fdr_cutoff"] &
                                      pvm_ex_S2$B_amp > dfS1_2[i, "amp_cutoff"])
                tp <- sum(true_dr_genes & est_dr_genes)/nrow(pvm_ex_S2)
                fp <- sum(!true_dr_genes & est_dr_genes)/nrow(pvm_ex_S2)
                return(c(TP = tp, FP = fp))
              }, FUN.VALUE = numeric(2))
dfS1_2 <- cbind(dfS1_2, t(roc))

fS1_2 <- ggplot(dfS1_2) + 
  geom_area(aes_(x=~fdr_cutoff, y=~TP+FP), color="black", fill = "white") +
  geom_area(aes_(x=~fdr_cutoff, y=~TP), fill = "grey70", alpha = 0.75, position ="identity") + 
  facet_wrap(~factor(amp_cutoff), nrow = 1,
             labeller = as_labeller(setNames(paste0('~A >', amp_choices), amp_choices), label_parsed)) +
  geom_hline(aes_(yintercept=~sum(true_dr_genes)/nrow(pvm_ex_S2)), linetype = "twodash") +
  xlab("false discovery cutoff for rhythms") + ylab("DiffR transcripts") + 
  scale_x_continuous(trans = "log10", expand = c(0,0), breaks = c(0.01, 0.05, 0.2)) +
  coord_cartesian(xlim = c(0.005,0.25)) + scale_y_continuous(labels = scales::percent)

fS1_1 + fS1_2 + plot_layout(nrow=1, widths = c(1, 2)) + plot_annotation(tag_levels = 'A')
```

* on the subsampled data, the p-values are larger resulting is even over estimates for less conservative thresholds.
* for strigent cutoffs, the number of DR genes is underestimated with only a small fraction of DR genes.

## Analysis of datasets

### High-fat diet

- Very few number of diff-rhythmic genes ~ 100
```{r nutr_challenge_1, fig.align="center"}
se <- readRDS("data/GSE52333.RDS")

exp_des_nutr_1 <- colData(se) %>% as.data.frame %>% tibble() %>%
                 dplyr::select(!!c("harvest.timepoint.ch1", "diet.ch1")) %>%
                 mutate(time = as.numeric(sub("ZT", "", harvest.timepoint.ch1)),
                        group = fct_recode(diet.ch1, HFD="high fat diet", control="normal chow"),
                        group = fct_relevel(group, "control"))

y_nutr_1 <- assay(se, "expr")

group_ids <- levels(exp_des_nutr_1$group)
keep <- filter_genes(y_nutr_1, exp_des_nutr_1$group, min.expr = 5)

y_nutr_1 <- y_nutr_1[keep, ]

results_emahan_dodr <- compareRhythms(y_nutr_1, exp_des_nutr_1, method = "dodr", 
                                      just_classify = FALSE)
results_emahan <- compareRhythms(y_nutr_1, exp_des_nutr_1, method = "limma", 
                            just_classify = FALSE) %>% 
             inner_join(as.data.frame(rowData(se))[c(1,2,3)], 
                        by=c("id"="ensembl_gene_id"))
results_emahan_modsel <- compareRhythms(y_nutr_1, exp_des_nutr_1, method = "mod_sel", 
                                      just_classify = FALSE) %>%
            inner_join(as.data.frame(rowData(se))[c(1,2,3)], 
                        by=c("id"="ensembl_gene_id"))


compare_res <- inner_join(results_emahan_dodr, results_emahan, by="id")
df1 <- with(compare_res, table(diff_rhythmic.x, diff_rhythmic.y))
tt <- ttheme_default(colhead=list(fg_params=list(fontface="bold")),
  rowhead=list(fg_params=list(fontface="bold"), bg_params=list(fill="grey80")))
grid.table(df1, rows=c("not DR", "DR"), cols=c("not DR", "DR"), theme=tt)

# print(length(union(setdiff(results_emahan_dodr$id, results_emahan$id), setdiff(results_emahan$id, results_emahan_dodr$id)))/length(union(results_emahan_dodr$id, results_emahan$id))) # due to higher power of RAIN

df4_1 <- bind_rows(data.frame(method="DODR", fdr=results_emahan_dodr$adj_p_val_dodr), 
          data.frame(method="limma", fdr=results_emahan$adj_p_val_DR)) %>%
          group_by(method) %>%
          mutate(len=length(fdr))

f4_1 <- ggplot(data = df4_1) + stat_ecdf(aes(x=fdr, len=len, y=..y..*len, 
                                        color=factor(method)), geom = "step", size=0.75, pad=FALSE) +
     geom_linerange(aes(xmin=0.12, xmax=0.25, 
                    y=sum(results_emahan_modsel$category %in% c("gain", "loss", "change"))), 
                color=colorblind_pal()(3)[3], size=0.3) +
     xlab("false discovery rate") + ylab("number of DiffR transcripts") + 
     coord_cartesian(xlim = c(0,0.5), ylim = c(0, 1300)) + scale_color_colorblind() +
     theme(legend.position = "none")

df4_1 %<>% filter(row_number() == 1) %>%
      dplyr::select(method, len) %>%
      bind_rows(data.frame(method = "model sel.", 
                           len = sum(results_emahan_modsel$category!="arrhy"))) 

f4_1 <- f4_1 + geom_linerange(aes(y=len, color=method), xmin=0, xmax=0.1, 
                   data = df4_1, size=0.5) +
  geom_text(aes(y=len, label=method), x=0.06, data=df4_1, hjust=0, vjust=-0.3, size=9*0.35)


```

```{r nutr_challenge_2}
se <- readRDS("data/PRJNA428303.RDS") %>%
      subset(select = colnames(.) != "SRR6436173")

exp_des_nutr_2 <- colData(se) %>% as.data.frame %>% tibble %>%
                 dplyr::select(!!c("time", "group")) %>%
                 mutate(time = as.numeric(sub("ZT", "", time)),
                        group = as_factor(group))
                 
y_nutr_2 <- assay(se, "countsFromAbundance")

group_ids <- levels(exp_des_nutr_2$group)
keep <- filter_genes(y_nutr_2, group=exp_des_nutr_2$group, min.count = 5, type="rnaseq")

y_nutr_2 <- y_nutr_2[keep, ]

results_quag <- compareRhythms(y_nutr_2, exp_des_nutr_2, method = "voom", 
                           just_classify = FALSE, outliers = TRUE, robust = TRUE) %>%
            inner_join(as.data.frame(rowData(se))[c(1,2,3)], by=c("id"="ensembl_gene_id")) 


df4_2 <- data.frame(source = rep(c("Eckel-Mahan et al.", "Quagliarini et al."), each=2),
                 category = rep(c("DR", "ABR"), 2),
                 counts = c(table(results_emahan$category == "same"), table(results_quag$category == "same"))) %>%
      group_by(source) %>%
      mutate(fraction = counts/sum(counts)) %>%
      mutate(ymax = cumsum(fraction), ymin = c(0, head(ymax, n=-1)),
             labelPos = (ymax + ymin) / 2)

f4_2 <- ggplot(df4_2, aes(ymin=ymin, ymax=ymax, xmin=2, xmax=3, fill=category)) + 
  geom_rect(color="white") + coord_polar(theta = "y") + facet_wrap(~source, nrow = 1) +
  scale_fill_manual(values = c("grey20","grey90")) + 
  geom_text(x=2.5, aes(y=labelPos, label=counts, color=category), size=2.5, 
            fontface="bold", show.legend = TRUE) + scale_color_manual(values = c("grey90","grey20")) + 
     xlim(c(1,3)) + 
  theme(legend.position = "none", strip.background = element_blank(), 
                       axis.line = element_blank(), axis.ticks = element_blank(),
                       aspect.ratio = 1, axis.text = element_blank(), panel.grid.major.x = element_blank(),
                       panel.grid.major.y = element_blank(), axis.title = element_blank(),
                       title = element_blank(), panel.spacing.x = unit(0, "lines")) + ylab("") + xlab("")
```

```{r hfd_timecourses, fig.width=6.5, fig.height=3.5}
hfd_genes <- c("Nampt", "Nr1d1", "Per2", "Dbp", "Cry1", "Arntl", "Nr1d2", "Per1", "Rorc")
hfd_genes_ensembl <- filter(results_emahan, mgi_symbol %in% hfd_genes)

data_eckel <- y_nutr_1[hfd_genes_ensembl$id,] %>% 
           magrittr::set_colnames(paste(exp_des_nutr_1$group, exp_des_nutr_1$time, sep="_")) %>%
           magrittr::set_rownames(hfd_genes_ensembl$mgi_symbol) %>%
           data.frame %>% 
           rownames_to_column(var="id") %>% 
           gather(sample, value, -id) %>% 
           separate(sample, c("grp", "time"), "_", convert = TRUE) %>% 
           mutate(type = "eckel-mahan")

y_nutr_2_abund <- assay(se, "abundance") %>% {log2(. + 0.01)}

data_quag <- y_nutr_2_abund[hfd_genes_ensembl$id,] %>% 
           magrittr::set_colnames(paste(exp_des_nutr_2$group, exp_des_nutr_2$time, sep="_")) %>%
           magrittr::set_rownames(hfd_genes_ensembl$mgi_symbol) %>%
           data.frame %>% 
           rownames_to_column(var="id") %>% 
           gather(sample, value, -id) %>% 
           separate(sample, c("grp", "time"), "_", convert = TRUE) %>% 
           mutate(type = "quag")

df4_5 <- bind_rows(data_eckel, data_quag)

label_one <- function(labels) return(list(labels$id))
          
f4_5 <- ggplot(df4_5, aes(time, value, linetype=grp, color=type)) + 
  geom_point(size = 0.3) +
  geom_smooth(formula = 'y~x', method = "loess", size=0.4, se=FALSE, span=0.75) +
  facet_wrap(~ id + type, scales = "free_y", ncol = 6, labeller = label_one) + 
  theme_custom(base_size = 9) + scale_color_tableau(name=NULL, labels = c("Eckel-Mahan et al.", "Quagliarini et al.")) + scale_linetype_manual(values = c("dashed","solid"), name=NULL) + scale_y_continuous(breaks = scales::breaks_extended(n=4)) +
  theme(axis.text = element_text(size=8), axis.ticks.length = unit(4, "pt"), panel.spacing = unit(8, "pt")) +
  xlab("Zeitgeber time") + ylab(expression(log[2]~expression))

```



```{r nutr_phase_shift}
nutr_compare <- inner_join(results_emahan, results_quag, by=c("id", "entrezgene_id", "mgi_symbol")) %>%
               mutate(DR = diff_rhythmic.x | diff_rhythmic.y,
                      amp_change_1 = (HFD_amp.x - control_amp.x),
                      phase_change_1 = exp(1i*HFD_phase.x - 1i*control_phase.x),
                      amp_change_2 = (HFD_amp.y - control_amp.y),
                      phase_change_2 = exp(1i*HFD_phase.y - 1i*control_phase.y))

df4_3 <- filter(nutr_compare, DR) %>%
      dplyr::select(amp_change_1, phase_change_1, amp_change_2, phase_change_2, id) %>%
      gather(var, value, -id) %>%
      separate(var, c("var", "junk", "source")) %>% dplyr::select(-junk) %>%
      group_by(id, source) %>%
      spread(var, value)

f4_3 <- ggplot(data=df4_3) +
  geom_vline(xintercept = c(-1,1), linetype="dashed", color="grey80") +
  geom_vline(xintercept = 0, color = "grey80") +
  geom_hline(yintercept = seq(-2*pi/3, pi, pi/3), color="grey80") +
  geom_point(aes(x=Re(amp), y = Arg(phase), color=source), size=0.4) +
  facet_wrap(~source, nrow=1, labeller = as_labeller(c(`2` = "Quagliarini et al.", `1` = "Eckel-Mahan et al."))) + 
  theme(aspect.ratio = 1) + xlab("") + ylab("") + 
  geom_text(data=data.frame(x=seq(-1,1), y=pi/2), aes(x=x, y=y, label=x), nudge_x = -0.2, size=9*0.35) + 
  scale_x_continuous(breaks=seq(-2,1), limits = c(-2,1), expand = c(0.1,0)) + 
  scale_y_continuous(breaks = seq(-2*pi/3, pi, pi/3), limits=c(-pi,pi), 
                     labels = ((12/pi*seq(-2*pi/3, pi, pi/3) + 12) %% 24) - 12, ) +
  theme(aspect.ratio = 1, legend.position = "none", axis.line = element_blank(), 
        axis.ticks = element_blank(), 
        axis.text.x = element_text(size = 9),
        axis.text.y=element_blank(), panel.grid.major = element_blank()) + 
  coord_polar(theta="y", start=-pi) + scale_color_tableau()


```

```{r kegg}
kegg_1 <- enrichKEGG(results_emahan$entrezgene_id[results_emahan$diff_rhythmic], universe = as.character(results_emahan$entrezgene_id), organism = "mmu", keyType = "ncbi-geneid", pvalueCutoff = 1.0, qvalueCutoff = 1.0, minGSSize = 5)

kegg_2 <- enrichKEGG(results_quag$entrezgene_id[results_quag$diff_rhythmic], universe = as.character(results_quag$entrezgene_id), organism = "mmu", keyType = "ncbi-geneid", pvalueCutoff = 1.0, qvalueCutoff = 1.0, minGSSize = 5)

df4_4_1 <- head(kegg_1, n=5)[, 2, drop=FALSE] %>%
           mutate(Description = str_wrap(Description, width=25))
df4_4_2 <- head(kegg_2, n=5)[, 2, drop=FALSE] %>%
           mutate(Description = str_wrap(Description, width=25))

f4_4_1 <- tableGrob(df4_4_1, theme = ttheme_default(base_size = 8), cols = levels(df4_2$source)[1], rows = NULL)
f4_4_2 <- tableGrob(df4_4_2, theme = ttheme_default(base_size = 8), cols = levels(df4_2$source)[2], rows = NULL)
```

```{r reactome, eval=FALSE}
reactome_1 <- enricher(results$mgi_symbol[results_emahan$diff_rhythmic], universe = as.character(results_emahan$mgi_symbol), TERM2GENE = m_t2g_reactome, qvalueCutoff = 1.0, pAdjustMethod = "none", pvalueCutoff = 0.05)

reactome_2 <- enricher(results_quag$mgi_symbol[results_quag$diff_rhythmic], universe = as.character(results_quag$mgi_symbol), TERM2GENE = m_t2g, qvalueCutoff = 1.0, pAdjustMethod = "none", pvalueCutoff = 0.05)

```

```{r Figure4, fig.height=8, fig.width=6.5}
#grid.arrange(f4_1, f4_2, f4_3, layout_matrix=matrix(1:4, ncol=2, byrow=TRUE))

layout <- c(area(1,1), area(1,2), area(2,1,2,2), area(3,1), area(3,2))

f4_1 + f4_2 + f4_5 + gtable_cbind(f4_4_1, f4_4_2) + f4_3 + plot_layout(design = layout, widths = c(0.8,1), heights = c(0.66,1,0.66)) + plot_annotation(tag_levels = "A")

write.xlsx(list(results_emahan, results_quag, results_emahan_dodr, results_emahan_modsel), file="HFD_results.xlsx", 
           creator = "Bharath Ananthasubramaniam", 
           asTable = c(TRUE, TRUE, TRUE, TRUE),
           sheetName = c("Eckel-Mahan (limma)", "Quagliarini", "Eckel-Mahan (DODR)", "Eckel-Mahan (Model Sel.)"),
           borders = rep("surrounding",4))
```


```{r keto}
se <- readRDS("data/GSE87425.RDS") %>%
      subset(select = colnames(.) != "GSM2331169") %>%
      magrittr::extract(, .$tissue.ch1 == "LIVER") 

exp_des_keto <- colData(se) %>% as.data.frame %>% tibble %>%
                 dplyr::select(!!c("zt.ch1", "diet.ch1")) %>%
                 mutate(time = as.numeric(zt.ch1),
                        group = fct_recode(diet.ch1, keto="Ketogenic", control="Norma Chow"),
                        group = fct_relevel(group, "control"))

y_keto <- assay(se)

group_ids <- levels(exp_des_keto$group)
keep <- filter_genes(y_keto, exp_des_keto$group, min.expr = 5)

y_keto <- y_keto[keep, ]

results_keto <- compareRhythms(y_keto, exp_des_keto, method = "limma", 
                            just_classify = FALSE) %>% 
             inner_join(as.data.frame(rowData(se))[c(1,2,3)], by=c("id"="ensembl_gene_id"))
results_keto_dodr <- compareRhythms(y_keto, exp_des_keto, method = "dodr", 
                            just_classify = FALSE) %>% 
             inner_join(as.data.frame(rowData(se))[c(1,2,3)], by=c("id"="ensembl_gene_id")) 

results_keto_modsel <- compareRhythms(y_keto, exp_des_keto, method = "mod_sel", 
                            just_classify = FALSE) %>% 
             inner_join(as.data.frame(rowData(se))[c(1,2,3)], by=c("id"="ensembl_gene_id")) 
```


```{r keto_figs}
df5_1 <- bind_rows(data.frame(category = table(results_keto$category), method = "limma"),
                   data.frame(category = table(results_keto_dodr$category), method = "DODR"),
                   data.frame(category = table(results_keto_modsel$category), method = "model sel.")) %>%
         rename(category = category.Var1, freq = category.Freq) %>%
         dplyr::filter(category != "arrhy")

f5_1 <- ggplot(df5_1) +
  geom_col(aes(x=category, y=freq, fill=method), position = "dodge") + scale_fill_colorblind() +
  # annotate("text", x = 3, y = c(950, 800), label= c("A = control", "B = ketogenic"), hjust = 0, size = 0.35*9) +
  theme(legend.title = element_blank(), legend.key.size = unit(0.75,"line"),
        legend.justification = "left", legend.box.margin = margin(-5,-5,-5,-20)) + ylab("number of transcripts") + xlab("")


df5_2 <- results_keto %>%
         filter(diff_rhythmic) %>%
                 mutate(amp_change = (keto_amp - control_amp),
                      phase_change = exp(1i*keto_phase - 1i*control_phase)) %>%
         filter(category == "change")

f5_2 <- ggplot(data=df5_2) +
  geom_vline(xintercept = c(-1,1), linetype="dashed", color="grey80") +
  geom_vline(xintercept = 0, color = "grey80") +
  geom_hline(yintercept = seq(-2*pi/3, pi, pi/3), color="grey80") +
  geom_point(aes(x=amp_change, y = Arg(phase_change)), size=0.4) +
  theme(aspect.ratio = 1) + xlab("") + ylab("")  +
  geom_text(data=data.frame(x=seq(-1,1), y=pi/2), aes(x=x, y=y, label=x), nudge_x = -0.2, size=9*0.35) + 
  scale_x_continuous(breaks=seq(-2,1), limits = c(-2,1), expand = c(0.1,0)) + 
  scale_y_continuous(breaks = seq(-2*pi/3, pi, pi/3), limits=c(-pi,pi), 
                     labels = ((12/pi*seq(-2*pi/3, pi, pi/3) + 12) %% 24) - 12) +
  theme(aspect.ratio = 1, axis.line = element_blank(), 
        axis.ticks = element_blank(),
        axis.text.x = element_text(size = 9),
        axis.text.y=element_blank(), panel.grid.major = element_blank(),
        panel.border = element_blank()) + 
  coord_polar(theta="y", start=-pi)

```


```{r Figure5, fig.height=4, fig.width=6.5}
kegg_keto <- enrichKEGG(results_keto$entrezgene_id[results_keto$diff_rhythmic], universe = as.character(results_keto$entrezgene_id), organism = "mmu", keyType = "ncbi-geneid", pvalueCutoff = 1.0, qvalueCutoff = 1.0, minGSSize = 5)

hallmark_keto <- enricher(results_keto$mgi_symbol[results_keto$diff_rhythmic], universe = as.character(results_keto$mgi_symbol), TERM2GENE = m_t2g_hallmark, qvalueCutoff = 1.0, pAdjustMethod = "none")

df5_3_1 <- head(kegg_keto, n=10) %>%
           filter(qvalue<0.05) %>%
           mutate(Description = str_wrap(Description, width=18),
                  qvalue = scales::scientific(qvalue, digits = 3)) %>%
           dplyr::select(Description, qvalue)
df5_3_2 <- head(hallmark_keto, n=10) %>%
           filter(qvalue<0.05) %>%
           mutate(Description = gsub("_", " ", sub("HALLMARK","",Description)),
                  Description = str_wrap(str_to_sentence(Description), width=18),
                  qvalue = scales::scientific(qvalue, accuracy = 0.001)) %>%
           dplyr::select(Description, qvalue)

f5_3_1 <- tableGrob(df5_3_1, theme = ttheme_default(base_size = 8), cols = c("KEGG", "qvalue"), rows = NULL)
f5_3_2 <- tableGrob(df5_3_2, theme = ttheme_default(base_size = 8), cols = c("Hallmark (MSigDB)", "qvalue"), rows = NULL)


ifng_genes_all <- unlist(strsplit("Irf9/Fcgr1/Cd274/Dhx58/Cmpk2/Rsad2/Mx2/Eif2ak2/Fas/Irf7/Ifih1/Nmi/Zbp1/Helz2/Samhd1/Adar/Ifi44/Gbp3/Cxcl9/Usp18/Trim21/Oas3/Oas2/Rtp4/Lgals3bp/Parp14/Cxcl10/Pml/Tap1/Ddx60/Parp12/Isg20/Znfx1/Stat2/Ddx58/Xaf1/Oasl1/Trafd1/Ifit2/Samd9l/Rnf213/Ifit3", split="/"))

ifng_genes <- c("Irf9","Gbp3", "Rsad2", "Ifit3", "Dhx58", "Irf7")

ifng_genes_ensembl <- filter(results_keto, mgi_symbol %in% ifng_genes)

df5_4 <- y_keto[ifng_genes_ensembl$id,] %>% 
           magrittr::set_colnames(paste(exp_des_keto$group, exp_des_keto$time, sep="_")) %>%
           magrittr::set_rownames(ifng_genes_ensembl$mgi_symbol) %>%
           data.frame %>% 
           rownames_to_column(var="id") %>% 
           gather(sample, value, -id) %>% 
           separate(sample, c("grp", "time"), "_", convert = TRUE)

f5_4 <- ggplot(df5_4, aes(time, value, linetype=grp)) + 
  geom_point(size = 0.3) +
  geom_smooth(formula = 'y~x', method = "loess", size=0.4, se=FALSE, span=0.75, color="black") +
  facet_wrap(~ id, scales = "free_y", ncol = 6) + 
  theme_custom(base_size = 9) + scale_linetype_manual(values = c("dashed","solid"), name=NULL, labels=c("control","ketogenic")) + scale_y_continuous(breaks = scales::breaks_extended(n=4)) +
  theme(axis.text = element_text(size=8), axis.ticks.length = unit(4, "pt"), 
        panel.spacing.x = unit(8, "pt"), legend.box.margin = margin(b =-5), 
        legend.margin = margin(0,0,0,0), legend.title = element_blank()) +
  xlab("Zeitgeber time") + ylab(expression(log[2]~expression))

layout <- c(area(1,1), area(1,2), area(1,3), area(2,1,2,3))

f5_1 + f5_2 + gtable_rbind(f5_3_1, f5_3_2) + f5_4 + plot_layout(design = layout, heights = c(1,0.3), widths = c(0.9,0.8,1)) + plot_annotation(tag_levels = "A")

```


```{r intestine}
se <- readRDS("data/GSE87425.RDS") %>%
      magrittr::extract(, .$tissue.ch1 == "INTESTINE (IEC)") 

exp_des_keto_iec <- colData(se) %>% as.data.frame %>% tibble %>%
                 dplyr::select(!!c("zt.ch1", "diet.ch1")) %>%
                 mutate(time = as.numeric(zt.ch1),
                        group = fct_recode(diet.ch1, keto="Ketogenic", control="Norma Chow"),
                        group = fct_relevel(group, "control"))

y_keto_iec <- assay(se)

group_ids <- levels(exp_des_keto_iec$group)
keep <- apply(y_keto_iec, 1, function(x) median(x[exp_des_keto_iec$group == group_ids[1]])>5 | 
                                median(x[exp_des_keto_iec$group == group_ids[2]])>5 )

y_keto_iec <- y_keto_iec[keep, ]

results_keto_iec <- compareRhythms(y_keto_iec, exp_des_keto_iec, method = "limma", 
                            just_classify = FALSE, compare_fdr = 0.05) %>% 
             inner_join(as.data.frame(rowData(se))[c(1,2,3)], by=c("id"="ensembl_gene_id")) 

results_keto_iec_dodr <- compareRhythms(y_keto_iec, exp_des_keto_iec, method = "dodr", 
                            just_classify = FALSE, compare_fdr = 0.05) %>% 
             inner_join(as.data.frame(rowData(se))[c(1,2,3)], by=c("id"="ensembl_gene_id")) 
results_keto_iec_modsel <- compareRhythms(y_keto_iec, exp_des_keto_iec, method = "mod_sel", 
                            just_classify = FALSE) %>% 
             inner_join(as.data.frame(rowData(se))[c(1,2,3)], by=c("id"="ensembl_gene_id")) 
```

```{r FigureS4, fig.height=4.5, fig.width=6.5}
kegg_iec <- enrichKEGG(results_keto_iec$entrezgene_id[results_keto_iec$diff_rhythmic], universe = as.character(results_keto_iec$entrezgene_id), organism = "mmu", keyType = "ncbi-geneid", pvalueCutoff = 1.0, qvalueCutoff = 1.0, minGSSize = 5)

hallmark_iec <- enricher(results_keto_iec$mgi_symbol[results_keto_iec$diff_rhythmic], universe = as.character(results_keto_iec$mgi_symbol), TERM2GENE = m_t2g_hallmark)

dfS4_1 <- bind_rows(data.frame(category = table(results_keto_iec$category), method = "limma"),
                   data.frame(category = table(results_keto_iec_dodr$category), method = "DODR"),
                   data.frame(category = table(results_keto_iec_modsel$category), method = "model sel.")) %>%
         rename(category = category.Var1, freq = category.Freq) %>%
         dplyr::filter(category != "arrhy")

fS4_1 <- ggplot(dfS4_1) +
  geom_col(aes(x=category, y=freq, fill=method), position = "dodge") + scale_fill_colorblind() +
  # annotate("text", x = 3, y = c(950, 800), label= c("A = control", "B = ketogenic"), hjust = 0, size = 0.35*9) +
  theme(legend.title = element_blank(), legend.key.size = unit(0.75,"line"),
        legend.justification = "left", legend.box.margin = margin(-5,-5,-5,-20)) + ylab("number of transcripts") + xlab("")

dfS4_2 <- results_keto_iec %>%
         filter(diff_rhythmic) %>%
                 mutate(amp_change = (keto_amp - control_amp),
                      phase_change = exp(1i*keto_phase - 1i*control_phase)) %>%
         filter(category == "change")

fS4_2 <- ggplot(data=dfS4_2) +
  geom_vline(xintercept = c(-1,1), linetype="dashed", color="grey80") +
  geom_vline(xintercept = 0, color = "grey80") +
  geom_hline(yintercept = seq(-2*pi/3, pi, pi/3), color="grey80") +
  geom_point(aes(x=amp_change, y = Arg(phase_change)), size=0.5) +
  scale_shape_manual(values = c(0, 3, 20)) +
  theme(aspect.ratio = 1) + xlab("") + ylab("") +
  geom_text(data=data.frame(x=seq(-1,1), y=pi/2), aes(x=x, y=y, label=x), nudge_x = -0.2, size=9*0.35) + 
  scale_x_continuous(breaks=seq(-2,1), limits = c(-2,1), expand = c(0.1,0)) + 
  scale_y_continuous(breaks = seq(-2*pi/3, pi, pi/3), limits=c(-pi,pi), 
                     labels = ((12/pi*seq(-2*pi/3, pi, pi/3) + 12) %% 24) - 12, ) +
  theme(aspect.ratio = 1, legend.position = "top", axis.line = element_blank(), 
        axis.ticks = element_blank(), legend.title = element_blank(),
        axis.text.x = element_text(size = 9), legend.key.size = unit(1, "line"),
        axis.text.y=element_blank(), panel.grid.major = element_blank()) + 
  coord_polar(theta="y", start=-pi)

dfS4_3_1 <- head(kegg_iec, n=3) %>%
           mutate(Description = str_wrap(Description, width=20),
                  qvalue = scales::comma(qvalue, accuracy = 0.001)) %>%
           dplyr::select(Description, qvalue)
dfS4_3_2 <- head(hallmark_iec, n=10) %>%
           filter(qvalue<0.05) %>%
           mutate(Description = gsub("_", " ", sub("HALLMARK","",Description)),
                  Description = str_wrap(str_to_sentence(Description), width=20),
                  qvalue = scales::scientific(qvalue, digits = 3)) %>%
           dplyr::select(Description, qvalue)

fS4_3_1 <- tableGrob(dfS4_3_1, theme = ttheme_default(base_size = 8), cols = c("KEGG", "qvalue"), rows = NULL)
fS4_3_2 <- tableGrob(dfS4_3_2, theme = ttheme_default(base_size = 8), cols = c("Hallmark genes(MSigDB)", "qvalue"), rows = NULL)

keto_iec_genes <- c("Nr1d1", "Nr1d2", "Tef", "Npas2", "Arntl", "Per2", "Nfil3", "Dbp", "Ppara", "Pparg", "Cyp51", "Ppargc1a")

keto_iec_ensembl <- filter(results_keto_iec, mgi_symbol %in% keto_iec_genes)

dfS4_4 <- y_keto_iec[keto_iec_ensembl$id,] %>% 
           magrittr::set_colnames(paste(exp_des_keto_iec$group, exp_des_keto_iec$time, sep="_")) %>%
           magrittr::set_rownames(keto_iec_ensembl$mgi_symbol) %>%
           data.frame %>% 
           rownames_to_column(var="id") %>% 
           gather(sample, value, -id) %>% 
           separate(sample, c("grp", "time"), "_", convert = TRUE)

fS4_4 <- ggplot(dfS4_4, aes(time, value, linetype=grp)) + 
  geom_point(size = 0.3) +
  geom_smooth(formula = 'y~x', method = "loess", size=0.4, se=FALSE, span=0.75, color="black") +
  facet_wrap(~ id, scales = "free_y", ncol = 6) + 
  theme_custom(base_size = 9) + scale_linetype_manual(values = c("dashed","solid"), name=NULL, labels=c("control","ketogenic")) + scale_y_continuous(breaks = scales::breaks_extended(n=4)) +
  theme(axis.text = element_text(size=8), axis.ticks.length = unit(4, "pt"), 
        panel.spacing.x = unit(8, "pt"), legend.box.margin = margin(-5,-5,-5,-5), 
        legend.margin = margin(0,0,0,0), legend.title = element_blank()) +
  xlab("Zeitgeber time") + ylab(expression(log[2]~expression))

write.xlsx(list(results_keto, results_keto_dodr, results_keto_modsel, results_keto_iec, results_keto_iec_dodr, results_keto_iec_modsel), file="KD_results.xlsx", 
           creator = "Bharath Ananthasubramaniam", 
           asTable = rep(TRUE, 6),
           sheetName = c("Liver (limma)", "Liver (DODR)", "Liver (Model sel.)", "IEC (limma)", "IEC (DODR)", "IEC (Model sel.)"),
           borders = rep("surrounding",6))


layout <- c(area(1,1), area(1,2), area(1,3), area(2,1,2,3))

fS4_1 + fS4_2 + gtable_rbind(fS4_3_1, fS4_3_2) + fS4_4 + plot_layout(design = layout, heights = c(1,0.9), widths = c(0.9,0.8,1)) + plot_annotation(tag_levels = "A")

```

```{r pei}
se <- readRDS("data/PRJNA449625.RDS") 

exp_des_pei <- colData(se) %>% as.data.frame %>% tibble %>%
                 dplyr::select(!!c("treatment", "Genotype", "replicate")) %>%
                 mutate(time = as.numeric(gsub("CT", "", treatment)),
                        group = fct_inorder(Genotype),
                        group = fct_relevel(group, "WT"),
                        batch = factor(make.names(replicate)))

y_pei <- assay(se, "countsFromAbundance")

group_ids <- levels(exp_des_pei$group)
keep <- filter_genes(y_pei, group=exp_des_pei$group)  #rownames(y_nutr_2) %in% rownames(y_nutr_1)

y_pei <- y_pei[keep, ]

results_pei <- compareRhythms(y_pei, exp_des_pei, method = "voom", 
                           just_classify = FALSE, compare_fdr = 0.05, outliers = TRUE) %>%
            inner_join(as.data.frame(rowData(se))[c(1,2,3)], by=c("id"="ensembl_gene_id"))

counts <- assay(se, "counts")
lengths <- assay(se, "lengths")

results_pei_deseq <- compareRhythms(counts[keep,], exp_design = exp_des_pei, 
                                    lengths = lengths[keep,], method = "deseq2", just_classify = FALSE) %>%
                     inner_join(as.data.frame(rowData(se))[c(1,2,3)], by=c("id"="ensembl_gene_id"))

df6_1 <- bind_rows(data.frame(category = table(results_pei$category), method = "voom"),
                   data.frame(category = table(results_pei_deseq$category), method = "DESeq2")) %>%
         rename(category = category.Var1, freq = category.Freq)

f6_1 <- ggplot(df6_1) +
  geom_col(aes(x=category, y=freq, fill=method), position = "dodge", width=0.5) + scale_fill_colorblind() +
  scale_y_log10() +
  # annotate("text", x = 3, y = c(2000, 600), label= c("A = WT", "B = p66KO"), hjust = 0, size = 0.35*9) +
  theme(legend.title = element_blank(), legend.key.size = unit(0.75,"line"),
        legend.justification = "left", legend.box.margin=margin(b=-5), legend.margin = margin()) + 
  ylab("number of transcripts") + xlab("")

df6_2 <- results_pei_deseq %>%
         filter(diff_rhythmic) %>%
                 mutate(amp_change = (P66KO_amp - WT_amp),
                      phase_change = exp(1i*P66KO_phase - 1i*WT_phase)) %>%
         filter(category == "change")

f6_2 <- ggplot(data=df6_2) +
  geom_vline(xintercept = c(-1,1), linetype="dashed", color="grey80") +
  geom_vline(xintercept = 0, color = "grey80") +
  geom_hline(yintercept = seq(-2*pi/3, pi, pi/3), color="grey80") +
  geom_point(aes(x=amp_change, y = Arg(phase_change)), size=0.2, alpha=0.8) +
  theme(aspect.ratio = 1) + xlab("") + ylab("") +
  scale_shape_manual(values = c(0, 3, 20)) +
  geom_text(data=data.frame(x=seq(-1,1), y=pi/2), aes(x=x, y=y, label=x), nudge_x = -0.2, size=9*0.35) + 
  scale_x_continuous(breaks=seq(-2,1), limits = c(-2,1), expand = c(0.1,0)) + 
  scale_y_continuous(breaks = seq(-2*pi/3, pi, pi/3), limits=c(-pi,pi), 
                     labels = ((12/pi*seq(-2*pi/3, pi, pi/3) + 12) %% 24) - 12, ) +
  theme(axis.line = element_blank(), axis.ticks = element_blank(),
        axis.text.x = element_text(size = 9), axis.text.y=element_blank(), 
        panel.grid.major = element_blank(), plot.margin = margin(), 
        axis.title = element_blank(), plot.background = element_blank()) + 
  coord_polar(theta="y", start=-pi)


```

```{r Figure6, fig.height=4.5, fig.width=6.5}
go_pei <- enrichGO(results_pei_deseq$entrezgene_id[results_pei_deseq$diff_rhythmic], org.Mm.eg.db, ont = "BP", universe = as.character(results_pei$entrezgene_id), qvalueCutoff = 0.5)

kegg_pei <- enrichKEGG(results_pei_deseq$entrezgene_id[results_pei_deseq$diff_rhythmic], universe = as.character(results_pei_deseq$entrezgene_id), organism = "mmu", keyType = "ncbi-geneid", pvalueCutoff = 1.0, qvalueCutoff = 1.0, minGSSize = 5)

hallmark_pei <- enricher(results_pei_deseq$mgi_symbol[results_pei_deseq$diff_rhythmic], universe = as.character(results_pei_deseq$mgi_symbol), TERM2GENE = m_t2g_hallmark)

df6_3_1 <- head(go_pei, n=3) %>%
           mutate(Description = str_wrap(Description, width=20),
                  qvalue = scales::scientific(qvalue, digits = 3)) %>%
           dplyr::select(Description, qvalue)
df6_3_2 <- head(hallmark_pei, n=10) %>%
           filter(qvalue<0.05) %>%
           mutate(Description = gsub("_", " ", sub("HALLMARK","",Description)),
                  Description = str_wrap(str_to_sentence(Description), width=20),
                  qvalue = scales::scientific(qvalue, digits = 3)) %>%
           dplyr::select(Description, qvalue)

f6_3_1 <- tableGrob(df6_3_1, theme = ttheme_default(base_size = 8), cols = c("GO (BP)", "qvalue"), rows = NULL)
f6_3_2 <- tableGrob(df6_3_2, theme = ttheme_default(base_size = 8), cols = c("Hallmark genes \n (MSigDB)", "qvalue"), rows = NULL)

y_pei_abund <- assay(se, "abundance") %>% {log2(. + 0.01)}

y_pei_abund <- limma::removeBatchEffect(y_pei_abund, batch = exp_des_pei$batch)

kras_genes <- unlist(strsplit("Sox9/G0s2/Dusp6/Flt4/Pecam1/Eng/Bmp2/Sparcl1/Plvap/Nr0b2/Cfhr1/Tspan7", split="/"))

kras_genes_ensembl <- filter(results_pei_deseq, mgi_symbol %in% kras_genes)

df6_4 <- y_pei_abund[kras_genes_ensembl$id,] %>% 
           magrittr::set_colnames(paste(exp_des_pei$group, exp_des_pei$time, sep="_")) %>%
           magrittr::set_rownames(kras_genes_ensembl$mgi_symbol) %>%
           data.frame %>% 
           rownames_to_column(var="id") %>% 
           gather(sample, value, -id) %>% 
           separate(sample, c("grp", "time"), "_", convert = TRUE) %>%
           mutate(grp = fct_relevel(grp, "WT"))

f6_4 <- ggplot(df6_4, aes(time, value, linetype=grp)) + 
  geom_point(size = 0.3) +
  geom_smooth(formula = 'y~x', method = "loess", size=0.4, se=FALSE, span=0.75, color="black") +
  facet_wrap(~ id, scales = "free_y", ncol = 6, labeller = label_one) + 
  theme_custom(base_size = 9)  + scale_linetype_manual(values = c("dashed","solid"), name=NULL, labels=c("control", expression(p66^Shc~KO))) + 
  scale_y_continuous(breaks = scales::breaks_extended(n=4)) +
  theme(axis.text = element_text(size=8), axis.ticks.length = unit(4, "pt"), panel.spacing = unit(8, "pt"), legend.box.margin = margin(b=-5, t=-10)) +
  xlab("Zeitgeber time") + ylab(expression(log[2]~expression))

layout <- c(area(1,1), area(1,2), area(1,3), area(2,1,2,3))

f6_1 + f6_2 + gtable_rbind(f6_3_1, f6_3_2) + f6_4 + plot_layout(design = layout, widths = c(1,0.9,1), heights = c(1,1)) + plot_annotation(tag_levels = "A")
```


